nix/offchain.nix

```diff
@@ -1,12 +1,463 @@
+                  Stringified path to the webpack `entrypoint` file.
+
+                  Added in: 2.1
+                '';
+
+                # NOTE: ideally, this would be a types.path, but it's easier to
+                # conform to CTL's types.
+                type = types.str;
+                default = "index.js";
+              };
+
+              browserRuntime = lib.mkOption {
+                description = ''
+                  Whether this bundle is being produced for a browser environment or
+                  not.
+
+                  Added in: 2.1
+                '';
+                type = types.bool;
+                default = true;
+              };
+
+              webpackConfig = lib.mkOption {
+                description = ''
+                  Stringified path to the Webpack config file to use.
+
+                  Added in: 2.1
+                '';
+                type = types.str;
+                default = "webpack.config.js";
+              };
+
+              bundledModuleName = lib.mkOption {
+                description = ''
+                  The name of the file containing the bundled JS module that
+                  `spago bundle-module` will produce.
+
+                  Added in: 2.1
+                '';
+                type = types.str;
+                default = "output.js";
+              };
+
+              enableCheck = lib.mkOption {
+                description = ''
+                  Whether to add a flake check testing that the bundle builds
+                  correctly.
+
+                  Added in: 2.1
+                '';
+                type = types.bool;
+                default = false;
+              };
+            };
+          };
+
+          testConfigs = types.submodule {
+            options = {
+              buildInputs = lib.mkOption {
+                type = types.listOf types.package;
+                description = ''
+                  Additional packages passed through to the `buildInputs` of
+                  the derivation.
+
+                  Added in: 2.1
+                '';
+                default = [ ];
+              };
+
+              testMain = lib.mkOption {
+                description = ''
+                  The name of the main Purescript module containing the test suite.
+
+                  Added in: 2.1
+                '';
+                type = purescriptModule;
+              };
+            };
+          };
+
+          runtime = types.submodule {
+            options = {
+              enableCtlServer = lib.mkOption {
+                description = ''
+                  Whether to enable or disable the CTL server (used to apply
+                  arguments to scripts and evaluate UPLC). Enabling this will
+                  also add the ctl-server overlay.
+
+                  Added in: 2.1
+                '';
+                type = types.bool;
+                default = false;
+              };
+
+              extraConfig = lib.mkOption {
```

Comment:
`extraConfig` can also be an overly-like function that takes the final value. https://github.com/Plutonomicon/cardano-transaction-lib/blob/a83dff3465f67c8d7b339bd5925a1440c5aab977/nix/runtime.nix#L104
This is quite useful and I think we need to allow setting extraConfig using a function

Reply:
The problem with using fixpoints like that is they're usually a nightmare to debug when something goes wrong. Could you give me an example of a config that would benefit from this change?

Reply:
If that's the case, it would be nice to add the wallet server option. AP is in the Liqwid-onchain repo.

Reply:
https://github.com/Liqwid-Labs/liqwid-offchain/pull/222

Reply:
I'd propose holding off on that until we actually merge it to off-chain. We can then iterate on this flake adding more scripts, tools, etc according to what we're using.

Reply:
Yeah, okay :+1: